### ISSUE TITLE
Support for lighting within instances

### ISSUE THREAD

**Issue by asluk** [PixarAnimationStudios/USD#1217]
If you open the attached usda in usdview built from the current dev branch, you'll see the effect of one sphere light that is under an xform. A second sphere light is under another xform marked instanceable, but its effect does not show up until you SetInstanceable(False) on its parent xform.

[sphereLightInstancingTest.zip](https://github.com/PixarAnimationStudios/USD/files/4657721/sphereLightInstancingTest.zip)

See usd-interest discussion at https://groups.google.com/d/msg/usd-interest/xROqlcv8Iqc/VB5jwBvPAwAJ

**Comment by jtran56** [comment]
Filed as internal issue #USD-6082.

**Comment by asluk** [comment]
There's a regression in v20.05, also reproducible in v20.08-rc4, where if you set the instanced Xform_01 in the sphereLightInstancing Test asset in the original report to be instanceable=false, usdview crashes

 	ntdll.dll!00007ff85c6bb0b9()	Unknown
 	ntdll.dll!00007ff85c6bb083()	Unknown
 	ntdll.dll!00007ff85c6c390e()	Unknown
 	ntdll.dll!00007ff85c6c3c1a()	Unknown
 	ntdll.dll!00007ff85c65ecb1()	Unknown
 	ntdll.dll!00007ff85c66ce62()	Unknown
 	ucrtbase.dll!00007ff85937c7eb()	Unknown
	[Inline Frame] usdImaging.dll!std::_Deallocate(void *) Line 207	C++
 	[Inline Frame] usdImaging.dll!std::_Default_allocator_traits<std::allocator<std::_Tree_node<pxrInternal_v0_20__pxrReserved__::SdfPath,void *> > >::deallocate(std::allocator<std::_Tree_node<pxrInternal_v0_20__pxrReserved__::SdfPath,void *> > &) Line 873	C++
 	[Inline Frame] usdImaging.dll!std::_Tree_node<pxrInternal_v0_20__pxrReserved__::SdfPath,void *>::_Freenode0(std::allocator<std::_Tree_node<pxrInternal_v0_20__pxrReserved__::SdfPath,void *> > &) Line 414	C++
 	[Inline Frame] usdImaging.dll!std::_Tree_comp_alloc<std::_Tset_traits<pxrInternal_v0_20__pxrReserved__::SdfPath,std::less<pxrInternal_v0_20__pxrReserved__::SdfPath>,std::allocator<pxrInternal_v0_20__pxrReserved__::SdfPath>,0> >::_Freeheadnode(std::_Tree_node<pxrInternal_v0_20__pxrReserved__::SdfPath,void *> *) Line 947	C++
 	[Inline Frame] usdImaging.dll!std::_Tree_comp_alloc<std::_Tset_traits<pxrInternal_v0_20__pxrReserved__::SdfPath,std::less<pxrInternal_v0_20__pxrReserved__::SdfPath>,std::allocator<pxrInternal_v0_20__pxrReserved__::SdfPath>,0> >::{dtor}() Line 813	C++
 	[Inline Frame] usdImaging.dll!std::_Tree<std::_Tset_traits<pxrInternal_v0_20__pxrReserved__::SdfPath,std::less<pxrInternal_v0_20__pxrReserved__::SdfPath>,std::allocator<pxrInternal_v0_20__pxrReserved__::SdfPath>,0> >::{dtor}() Line 1169	C++
 	usdImaging.dll!pxrInternal_v0_20__pxrReserved__::UsdImagingInstanceAdapter::_InstancerData::~_InstancerData()	C++
 	usdImaging.dll!std::list<std::pair<pxrInternal_v0_20__pxrReserved__::SdfPath const ,pxrInternal_v0_20__pxrReserved__::UsdImagingInstanceAdapter::_InstancerData>,std::allocator<std::pair<pxrInternal_v0_20__pxrReserved__::SdfPath const ,pxrInternal_v0_20__pxrReserved__::UsdImagingInstanceAdapter::_InstancerData> > >::erase(std::_List_const_iterator<std::_List_val<std::_List_simple_types<std::pair<pxrInternal_v0_20__pxrReserved__::SdfPath const ,pxrInternal_v0_20__pxrReserved__::UsdImagingInstanceAdapter::_InstancerData> > > > _Where) Line 1351	C++
 	[Inline Frame] usdImaging.dll!std::_Hash<std::_Umap_traits<pxrInternal_v0_20__pxrReserved__::SdfPath,pxrInternal_v0_20__pxrReserved__::UsdImagingInstanceAdapter::_InstancerData,std::_Uhash_compare<pxrInternal_v0_20__pxrReserved__::SdfPath,pxrInternal_v0_20__pxrReserved__::SdfPath::Hash,std::equal_to<pxrInternal_v0_20__pxrReserved__::SdfPath> >,std::allocator<std::pair<pxrInternal_v0_20__pxrReserved__::SdfPath const ,pxrInternal_v0_20__pxrReserved__::UsdImagingInstanceAdapter::_InstancerData> >,0> >::erase(std::_List_const_iterator<std::_List_val<std::_List_simple_types<std::pair<pxrInternal_v0_20__pxrReserved__::SdfPath const ,pxrInternal_v0_20__pxrReserved__::UsdImagingInstanceAdapter::_InstancerData> > > >) Line 607	C++
 	usdImaging.dll!std::_Hash<std::_Umap_traits<pxrInternal_v0_20__pxrReserved__::SdfPath,pxrInternal_v0_20__pxrReserved__::UsdImagingInstanceAdapter::_InstancerData,std::_Uhash_compare<pxrInternal_v0_20__pxrReserved__::SdfPath,pxrInternal_v0_20__pxrReserved__::SdfPath::Hash,std::equal_to<pxrInternal_v0_20__pxrReserved__::SdfPath> >,std::allocator<std::pair<pxrInternal_v0_20__pxrReserved__::SdfPath const ,pxrInternal_v0_20__pxrReserved__::UsdImagingInstanceAdapter::_InstancerData> >,0> >::erase<std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<pxrInternal_v0_20__pxrReserved__::SdfPath const ,pxrInternal_v0_20__pxrReserved__::UsdImagingInstanceAdapter::_InstancerData> > > >,void>(std::_List_iterator<std::_List_val<std::_List_simple_types<std::pair<pxrInternal_v0_20__pxrReserved__::SdfPath const ,pxrInternal_v0_20__pxrReserved__::UsdImagingInstanceAdapter::_InstancerData> > > > _Plist) Line 599	C++
 	usdImaging.dll!pxrInternal_v0_20__pxrReserved__::UsdImagingInstanceAdapter::_ResyncInstancer(const pxrInternal_v0_20__pxrReserved__::SdfPath & instancerPath, pxrInternal_v0_20__pxrReserved__::UsdImagingIndexProxy * index, bool repopulate) Line 1799	C++
 	usdImaging.dll!pxrInternal_v0_20__pxrReserved__::UsdImagingInstanceAdapter::_ResyncPath(const pxrInternal_v0_20__pxrReserved__::SdfPath & cachePath, pxrInternal_v0_20__pxrReserved__::UsdImagingIndexProxy * index, bool reload) Line 1397	C++
 	usdImaging.dll!pxrInternal_v0_20__pxrReserved__::UsdImagingDelegate::_ResyncUsdPrim(const pxrInternal_v0_20__pxrReserved__::SdfPath & usdPath, pxrInternal_v0_20__pxrReserved__::UsdImagingIndexProxy * proxy, bool repopulateFromRoot, bool fromThread) Line 1343	C++
 	usdImaging.dll!pxrInternal_v0_20__pxrReserved__::UsdImagingDelegate::ApplyPendingUpdates() Line 1151	C++
 	usdImagingGL.dll!pxrInternal_v0_20__pxrReserved__::UsdImagingGLEngine::_PreSetTime(const pxrInternal_v0_20__pxrReserved__::UsdPrim & root, const pxrInternal_v0_20__pxrReserved__::UsdImagingGLRenderParams & params) Line 1294	C++
 	usdImagingGL.dll!pxrInternal_v0_20__pxrReserved__::UsdImagingGLEngine::PrepareBatch(const pxrInternal_v0_20__pxrReserved__::UsdPrim & root, const pxrInternal_v0_20__pxrReserved__::UsdImagingGLRenderParams & params) Line 235	C++
 	usdImagingGL.dll!pxrInternal_v0_20__pxrReserved__::UsdImagingGLEngine::Render(const pxrInternal_v0_20__pxrReserved__::UsdPrim & root, const pxrInternal_v0_20__pxrReserved__::UsdImagingGLRenderParams & params) Line 300	C++
 	[Inline Frame] _usdImagingGL.pyd!boost::python::detail::invoke(boost::python::detail::invoke_tag_<1,1>) Line 92	C++
 	[Inline Frame] _usdImagingGL.pyd!boost::python::detail::caller_arity<3>::impl<void (__cdecl pxrInternal_v0_20__pxrReserved__::UsdImagingGLEngine::*)(pxrInternal_v0_20__pxrReserved__::UsdPrim const &,pxrInternal_v0_20__pxrReserved__::UsdImagingGLRenderParams const &),boost::python::default_call_policies,boost::mpl::vector4<void,pxrInternal_v0_20__pxrReserved__::UsdImagingGLEngine &,pxrInternal_v0_20__pxrReserved__::UsdPrim const &,pxrInternal_v0_20__pxrReserved__::UsdImagingGLRenderParams const &> >::operator()(_object *) Line 216	C++
 	_usdImagingGL.pyd!boost::python::objects::caller_py_function_impl<boost::python::detail::caller<void (__cdecl pxrInternal_v0_20__pxrReserved__::UsdImagingGLEngine::*)(pxrInternal_v0_20__pxrReserved__::UsdPrim const &,pxrInternal_v0_20__pxrReserved__::UsdImagingGLRenderParams const &),boost::python::default_call_policies,boost::mpl::vector4<void,pxrInternal_v0_20__pxrReserved__::UsdImagingGLEngine &,pxrInternal_v0_20__pxrReserved__::UsdPrim const &,pxrInternal_v0_20__pxrReserved__::UsdImagingGLRenderParams const &> > >::operator()(_object * args, _object * kw) Line 38	C++
 	[External Code]	
 	tf.dll!pxrInternal_v0_20__pxrReserved__::Tf_ModuleProcessor::_InvokeWithErrorHandling::operator()(const boost::python::tuple & args, const boost::python::dict & kw) Line 181	C++
 	[Inline Frame] tf.dll!boost::python::detail::raw_dispatcher<pxrInternal_v0_20__pxrReserved__::Tf_ModuleProcessor::_InvokeWithErrorHandling>::operator()(_object *) Line 29	C++
 	tf.dll!boost::python::objects::full_py_function_impl<boost::python::detail::raw_dispatcher<pxrInternal_v0_20__pxrReserved__::Tf_ModuleProcessor::_InvokeWithErrorHandling>,boost::mpl::vector1<_object *> >::operator()(_object * args, _object * kw) Line 94	C++
 	[External Code]

**Comment by c64kernal** [comment]
Thanks for the report @asluk  -- mind filing a new issue for that? We'll try to get that fixed quicker than what's in this original issue.

**Comment by asluk** [comment]
Thanks @c64kernal -- just tried this again in dev and it doesn't crash. I think the issue is specific to our fork. Sorry for the noise!

**Comment by asluk** [comment]
Fixed the crash in our code-- was just a bad merge. Apologies again!

**Comment by c64kernal** [comment]
Great, glad to hear you're unblocked!

**Comment by pmolodo** [comment]
So, it seems that while Hydra should support instanced lights since 23.08, and PrMan since 23.11, Storm still doesn't have support:
https://forum.aousd.org/t/instanced-lights/980/6

Not sure if this ticket should be closed (since Hydra supports it), or updated to refer only to Storm support.

I did update the test scene @asluk included, as that one was old enough that it didn't use the current `inputs:` prefixes:

[sphereLightInstancingTest.v002.instancedLight.zip](https://github.com/user-attachments/files/16630615/sphereLightInstancingTest.v002.instancedLight.zip)

I also made 2 separate files:
- sphereLightInstancingTest.v002.instancedLight.usda
- sphereLightInstancingTest.v002.referencedLight.usda

...where the only difference is whether `instanceable = true`, to more easily compare expected vs instanced behavior.
